# compare_fasta
Analyse recall and precision of gene calls from [ggCaller](https://github.com/samhorsfield96/ggCaller) and [Prodigal](https://github.com/hyattpd/Prodigal).

## Requirements
- ```Biopython```

## Inputs
- ```--seq``` a FASTA of source sequences
- ```--genes``` a FASTA of known genes as nucleotide sequences
- ```--query``` a FASTA of genes as nucleotide sequences generated by either Prodigal (use ```-d``` option to generate mRNA sequences) or ggCaller
- ```--caller``` gene caller used (```prod``` for Prodigal, ```ggc``` for ggCaller)
- ```--min-size``` minimum size of genes called (default = 90 bp)
- ```--vers``` if ggCaller used, version in X.X format e.g. 0.1, 1.1, 1.2

## To run

If running with gene calls from ggCaller version 1.2:

```python compare_fasta.py --seq ref.fasta --genes genes.fasta --query query.fasta --caller ggc --vers 1.2```

If running with gene calls from Prodigal:

```python compare_fasta.py --seq ref.fasta --genes genes.fasta --query query.fasta --caller prod```
